pkg/costattribution/sample_tracker.go

```diff
@@ -0,0 +1,305 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package costattribution
+
+import (
+	"bytes"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+const sep = rune(0x80)
+
+type observation struct {
+	lastUpdate         atomic.Int64
+	receivedSample     atomic.Float64
+	discardedSampleMtx sync.Mutex
+	discardedSample    map[string]*atomic.Float64
+	totalDiscarded     atomic.Float64
+}
+
+type SampleTracker struct {
+	userID                     string
+	labels                     []string
+	maxCardinality             int
+	receivedSamplesAttribution *prometheus.Desc
+	discardedSampleAttribution *prometheus.Desc
+	overflowLabels             []string
+	observed                   map[string]*observation
+	observedMtx                sync.RWMutex
+	// overflowSince is also protected by observedMtx, it is set when the max cardinality is exceeded
+	overflowSince    time.Time
+	overflowCounter  observation
```

Comment:
Everything that is protected by a mutex should go under a mutex in a separate group, to make it obvious.

Reply:
addressed in this commit cb5512bd0f3e765422d0e19164de3c23369552f2 for both trackers.